### ISSUE TITLE
Question: How could I separate between between ontology, site and inferred data?

### ISSUE THREAD

**Issue by zoopp** [BrickSchema/py-brickschema#68]
Hey there, I want to get your opinion on something that I'm thinking of doing:

I want to do inference over datasets but would like to ideally separate boundaries between the used ontology data (i.e. brick, qudt etc.), the input site data (e.g. `:Sensor a brick:Temperature_Sensor`) and the inferred (output) data (e.g. :Sensor brick:hasTag brick:Temperature`). Since inference is always an additive process, this separation would easily allow me to change the site data (or ontology data) without worrying about the data that has been previously inferred (I can drop all inferred data and re-do inference).

I've been browsing the rdflib and brickschema source to get more familiar with them and I think that a way this might be doable is by reimplementing `brickschema.Graph.expand` on top of `rdflib.Dataset`. Separation could then be handled by adding a graph per each category of data (e.g. ontology data, site data, inferred data), doing inference on top of it all and storing the results in the inferred data graph.

Does this sound *reasonable*? Would something like this work? Is there a chance it might lead to potential complications (e.g. around querying for examples)? Is there maybe an easier way to do it?

**Comment by gtfierro** [comment]
Hi @zoopp 

This is an interesting idea, and a compelling workflow. Usually, I keep the Brick graph and my building graph in separate files, and then load them into the same in-memory graph to do ad-hoc queries. Sometimes, I will materialize the Brick + building + inferenced triples into a single graph and save that as a file that I can load into memory and perform queries on.  However, I think I like the way you have proposed better. I'm not sure if the way to do it is by reimplementing `brickschema.Graph.expand` on top of `rdflib.Dataset`, or if we should introduce a separate `brickschema.Dataset` on top of `rdflib.Dataset` that can be a little more opinionated: maybe it keeps a copy of Brick and a few different buildings and one could choose to access the pre-reasoned or the post-reasoned building graph either on a site-by-site basis, or by looking at the whole portfolio.

In short, this definitely sounds reasonable! What do you think of managing multiple buildings through an augmented `brickschema.Dataset` class?

**Comment by zoopp** [comment]
Hi @gtfierro,

Sorry for late reply, some things got in the way and I only noticed just now that you replied. The workflow that you describe at the beginning of the reply is very similar to what I was thinking to do initially, e.g., keep the raw site data and materialized data in separate datasets. I haven't fully explored this variant however I can easily spot some complexity around managing updates and tracking different files that would be easier to deal with with the other approach.

In my case, I find it easier to separate between sites on a dataset boundary (i.e. one dataset per site) however I don't see an issue with managing multiple sites through a single dataset (except maybe for longer query times when targeting individual sites?). In fact, this looks like a generalization of the data separation process that I'm after so things could work out regardless.

When it comes to running actual SPRQL queries I found that rdflib can be quite slow for complex queries so for now I plan to only use it for materialization and exporting the materialized data to a SPARQL server and using that for running queries (I'm currently evaluating Apache Fuseki). With that in mind, at least for now, I only plan to use py-brickschema to load raw data, materialize it and then export it to Fuseki for further consumption. Things would work out as long as I can somehow distinguish between raw and inferred data.

---

Various ideas popped into my mind while writing this reply and I've realised that in my case I don't really need to have this data separation handled with py-brickschema directly. Since I have full control over which graphs belonging to a dataset I write/query I can simply make sure to store the raw and materialized data in different graphs belonging to the same dataset (pretty obvious, right? :facepalm:).

Regardless, I'm open to further expand on this data separation idea if it looks like something worth adding to py-brickschema.

**Comment by gtfierro** [comment]
Agreed that RDFlib can be quite slow for complex queries. I've had some success with [pyoxigraph](https://pypi.org/project/pyoxigraph/) for faster results without having to reach into a whole other server.

The RDFlib philosophy towards managing many graphs has been, as per my understanding, to use the facilities of the Python language to roll whatever organization is most helpful to you. I think it would be helpful to develop some idioms, or perhaps a small library, for managing multiple Brick models, the ontology, and even materialized versions of the models. Ideally, this could help address some of the issues around how to effectively version Brick models and provide access to versions of models over time though a simple API. Not sure how much of this will get addressed in this issue, however :)

It was very helpful to learn about your preference for organizing these models. If you have any thoughts for how this project (or another project) could better manage Brick models over their lifecycle, please let me know!

**Comment by zoopp** [comment]
`pyoxigraph` looks interesting, I'll definitely look into it at some point.

I'm glad to hear that my ramblings are helpful :sweat_smile:. I'm still in a somewhat late early phase with figuring how to best organize this but I would be happy to keep in touch if you'd be interested in any specific points that you'd like to hear more about as I go. I can give you some background on what I'm working on to make this easier: essentially I'm developing a (private) platform that, among other things, has to deal with a lot of points coming from devices present in a lot of buildings and I'm trying to add some structure to it so that later I might be able to leverage this to do interesting things around navigation, discovery, visualization and troubleshooting.

Regarding the question itself, I found an answer that works for my usecase for now so I'm okay with having it closed and moving potential subsequent discussions on other points in different issues.

**Comment by gtfierro** [comment]
Sounds good! We can close for now; definitely let me know if you have any ideas for abstractions that py-brickschema could introduce in order to make tasks like that easier
